### ISSUE TITLE
Unique Constraints

### ISSUE THREAD

**Issue by jcbdev** [aws-amplify/amplify-category-api#286]

**Comment by nxia416** [comment]
+1

**Comment by acusti** [comment]
did the behavior of `attribute_exists` / `attribute_not_exists` dynamodb condition expressions change at some point? i was using `attribute_not_exists` in a `CreateItem` mutation (via `condition: { field: { attributeExists: false } }`), and i thought i had verified that it worked to ensure that the new item’s value for the specified field would be unique amongst all other items in the table, but i can verify that is not the case right now.

according to the [dynamodb docs](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.OperatorsAndFunctions.html) on comparison operators and functions, `attribute_exists` and `attribute_not_exists` are used to check whether an item has or doesn’t have an attribute. does the `VTL` version of that expression not mean the same thing?

@jcbdev do you know if your technique of maintaining a `UniqueConstraints` table with the id of each item being a hash of table + fieldname + value still works? or maybe it only works because it’s being used as the primary key, and primary keys enforce uniqueness, so it has nothing to do with the `attribute_not_exists(id)` condition? in which case i suppose that condition passes because the item hasn’t been created yet so it doesn’t have an id? and lastly, have you added a`Delete` item for the previous value of the unique field on `Update` and `Delete` operations? it would be great to see a more complete example of how i could implement this manually by modifying the VTL templates.

**Comment by Voyager-Two** [comment]
One easy option at the moment is to set unique fields inside primary key.
```graphql
id: ID! @primaryKey(sortKeyFields: ["username"])
```

**Comment by Voyager-Two** [comment]
This seems like major missing feature IMO.

**Comment by lewisdonovan** [comment]
Hi all. Just stumbled across this and think I might have found the answer. If you look at the generated schemas in AppSync, you'll see the schemas for the condition expression objects as well. For example, the schema to create a condition on a `String` input is:

```
input ModelStringInput {
	ne: String
	eq: String
	le: String
	lt: String
	ge: String
	gt: String
	contains: String
	notContains: String
	between: [String]
	beginsWith: String
	attributeExists: Boolean
	attributeType: ModelAttributeTypes
	size: ModelSizeInput
}
```

The first item in that schema `ne` means `not equal to`, which can be used to run a uniqueness comparison when putting an item. An example mutation would look like this:

```
mutation CreateTodoGraphQL($input: CreateTodoInput!, $condition: ModelTodoConditionInput) {
  createTodoGraphQL(input: $input, condition: $condition) {
    id
    name
    priority
    completedAt
  }
}

```

and the input variables would look like:

```
{
  "input": {
    "priority": "test",
    "name": "test",
    "completedAt": "2022-09-11T09:00:00.000Z"
  },
  "condition": { 
    "name": {
      "ne": "test"
    }
  }
}
```

If an item already exists with the name `test` then the condition expression would fail, and the item would not be added to the DB.

**Comment by Lookbid** [comment]
In my case, selling products via an auction with bidding, only one user should be able to bid, let's say $50. If another user places a $50, it should be ignored. Creating a new Lambda function seems like overkill. Would be nice to just have an @unique directive.

**Comment by Meerkov** [comment]
@lewisdonovan , isn't the condition specified on the client side? In other words, a client can cheat your system by removing that part of the request.

**Comment by lewisdonovan** [comment]
> @lewisdonovan , isn't the condition specified on the client side? In other words, a client can cheat your system by removing that part of the request.

Yes, absolutely. But AppSync can be run on the back-end too. You could create a custom mutation, pass `name` as a parameter (rather than a filter variable), map the mutation to a Lambda and run the filter logic there.

**Comment by gavmck** [comment]
@lewisdonovan I'm not sure the condition method works, just given it a spin and it still lets me create/update the user.
